{"nbformat":4,"nbformat_minor":0,"metadata":{"colab":{"provenance":[],"collapsed_sections":[],"toc_visible":true,"authorship_tag":"ABX9TyPXiHBJVZoMwDh7rvY/Gd/m"},"kernelspec":{"name":"python3","display_name":"Python 3"},"language_info":{"name":"python"}},"cells":[{"cell_type":"markdown","source":["# Assignment\n","\n","Using matplotlib/pyplot, implement a function that receives a pandas dataframe with three columns (chromosome, position, pvalue) and draws a manhattan plot with a result similar to this:\n","\n","![manhattan](https://github.com/ne1s0n/dataviz_python/raw/main/resources/proj01_manhattan_target.png)\n","\n","\n","\n"],"metadata":{"id":"CbxyzhEUKTu3"}},{"cell_type":"markdown","source":["# Data\n","\n","It is provided a GWAS result dataset originally included with the [Plink](https://github.com/ShujiaHuang/qmplot/blob/main/tests/data/gwas_plink_result.tsv) software, which I saved in the course repo for extra security. The file is:\n","\n","* tab separated\n","* without header\n","* 9224 rows (one per SNP)\n","* 12 columns:\n","  * `\"CHROM\", \"POS\", \"ID\", \"REF\", \"ALT\", \"A1\", \"TEST\", \"OBS_CT\", \"BETA\", \"SE\", \"T_STAT\", \"P\"`\n","* some rows have a missing value in the `P` column, and need to be removed (see [.dropna()](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.dropna.html#pandas.DataFrame.dropna))\n","\n","You may need a refresh on the [pandas.read_csv()](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.read_csv.html) function.\n","\n","The dataset is available at the following address:\n"],"metadata":{"id":"upCfe1plLGfi"}},{"cell_type":"code","execution_count":null,"metadata":{"id":"iVkIN23KKCNS"},"outputs":[],"source":["GWAS_RESULTS_URL = 'https://github.com/ne1s0n/dataviz_python/raw/main/resources/gwas_plink_result_cleaned.tsv'"]},{"cell_type":"markdown","source":["# Is there anybody NOT familiar with manhattan plots?\n","\n","Raise your hand!"],"metadata":{"id":"zs_vQ-plKnb0"}},{"cell_type":"markdown","source":["# Existing solutions\n","\n","Many software solutions are already available to create manhattan plots with more options that we will ever able to implement. There's no doubt that if you *need* to do a manhattan plot it's better to not reinvent the wheel and use something stable and published. The whole point of this project is to do meaningful exercise, though, not to create a competitive tool. \n","\n","That said, you may want to take a look to:\n","\n","* [manhattan plots with bioinfokit](https://www.reneshbedre.com/blog/manhattan-plot.html), a bioinformatics-oriented package\n","* [manhattan plots with plotly](https://plotly.com/python/manhattan-plot/), a generic visualization library with emphasis on interactivity\n","* [manhattan plots with qmplot](https://pythonawesome.com/a-python-package-for-creating-high-quality-manhattan-and-q-q-plots-from-gwas-results/) : a dedicated package for creating high-quality manhattan and Q-Q plots from GWAS results "],"metadata":{"id":"GsFfuFa1KXPY"}},{"cell_type":"markdown","source":["# That's it\n","\n","That's the assignment. Keep scrolling if you want to read and follow the subassignments (SA).\n","\n","<br><br><br><br><br><br>\n","<br><br><br><br><br><br>\n","<br><br><br><br><br><br>\n","<br><br><br><br><br><br>\n","<br><br><br><br><br><br>\n","<br><br><br><br><br><br>\n","<br><br><br><br><br><br>\n","<br><br><br><br><br><br>\n","<br><br><br><br><br><br>\n","<br><br><br><br><br><br>\n","<br><br><br><br><br><br>\n","<br><br><br><br><br><br>\n","<br><br><br><br><br><br>\n","<br><br><br><br><br><br>\n","<br><br><br><br><br><br>\n","<br><br><br><br><br><br>"],"metadata":{"id":"E76H-7mDqTZG"}},{"cell_type":"markdown","source":["# SA0) Read the data\n","\n","Read the data from the remote url saved in `GWAS_RESULTS_URL` variable, remembering that there's no file header but you should name the columns\n","\n","`\"CHROM\", \"POS\", \"ID\", \"REF\", \"ALT\", \"A1\", \"TEST\", \"OBS_CT\", \"BETA\", \"SE\", \"T_STAT\", \"P\"`\n","\n","Moreover remember to\n","\n","* remove lines with missing (`NA`) P values\n","* compute -log10(p)"],"metadata":{"id":"9Y6ZZgP7qxTj"}},{"cell_type":"code","source":["import pandas as pd\n","import numpy as np\n","\n","colnames = [\"CHROM\", \"POS\", \"ID\", \"REF\", \"ALT\", \"A1\", \"TEST\", \"OBS_CT\", \"BETA\", \"SE\", \"T_STAT\", \"P\"]\n","df = pd.read_csv(filepath_or_buffer = GWAS_RESULTS_URL, sep = '\\t', names = colnames)\n","print(df.shape)\n","df.dropna(inplace = True)\n","print(df.shape)\n","\n","#computing the -log10(p)\n","df['SCORE'] = -np.log10(df['P'])\n","print(df.shape)\n","\n","\n","df.head()"],"metadata":{"id":"_ftuI5DOHqS3"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["# SA1) A simpler problem: single chromosome manhattan plot \n","\n","At this point you should already be able to do a single chromosome plot."],"metadata":{"id":"aoLchaQpKQla"}},{"cell_type":"code","source":["import matplotlib.pyplot as plt\n","\n","#focusing on chromosome 1\n","df_ch1 = df[df['CHROM'] == 'chr01']\n","\n","plt.scatter(x = df_ch1['POS'], y=df_ch1['SCORE'])"],"metadata":{"id":"y6HN0_yXIGl3"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["# SA2) Sort the data by chromosome"],"metadata":{"id":"sFz-UZp9Nk9m"}},{"cell_type":"code","source":["df.sort_values(by = 'CHROM', inplace = True)"],"metadata":{"id":"_m6_uwLpNkbz"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["# SA3) Compute the length of each chromosome\n","\n","This will certainly come handy. Put the values in a variable.\n","\n"],"metadata":{"id":"U-ze98PdMSv9"}},{"cell_type":"code","source":["chroms = df.groupby('CHROM')['POS'].max()\n","print(chroms)"],"metadata":{"id":"ndqh7ZExME7M"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["chroms['chr01'] * 3"],"metadata":{"id":"axTXAI76MzwJ"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["# SA4) Create a new column SHIFT column\n","\n","For chrom 1, SHIFT should be zero\n","\n","For chrom 2, SHIFT should be the lenght of chrom 1\n","\n","For chrom 3, SHIFT should be the lenght of chrom 1 plus the length of chrom 2\n","\n","And so forth.\n"],"metadata":{"id":"gasxR3OKP0oa"}},{"cell_type":"code","source":["#first trial: let's just print the chromosome codes\n","for c in chroms.index:\n","  print(c)"],"metadata":{"id":"hsGTQWVsP6TS"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["#an improvement: let's compute the cumulative shift\n","shift = 0\n","for c in chroms.index:\n","  print('for ' + c + ' shift is ' + str(shift))\n","  shift = shift + chroms[c]"],"metadata":{"id":"exQcJTP-QShn"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["#the solution: let's compute the cumulative shift AND update the dataframe\n","\n","df['SHIFT'] = 0\n","\n","shift = 0\n","for c in chroms.index:\n","  print('for ' + c + ' shift is ' + str(shift))\n","\n","  sel = df['CHROM'] == c\n","  df.loc[sel, 'SHIFT'] = shift\n","\n","  shift = shift + chroms[c]\n"],"metadata":{"id":"v2Lt2S3LQ_ou"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["Taking a look to the SHIFT column"],"metadata":{"id":"en1ae9MYCoVq"}},{"cell_type":"code","source":["df[df['CHROM'] == 'chr01'].head()"],"metadata":{"id":"K5uWqWABRvYR"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["df[df['CHROM'] == 'chr02'].head()"],"metadata":{"id":"_tvgetPiCrIv"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["df[df['CHROM'] == 'chr03'].head()"],"metadata":{"id":"PFGUErEPCsCY"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["Testing that there is a single value for each chromosome"],"metadata":{"id":"e9JI8M4LCtt7"}},{"cell_type":"code","source":["df.groupby('CHROM')['SHIFT'].max() - df.groupby('CHROM')['SHIFT'].min()"],"metadata":{"id":"BI6KsgzIRzAo"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["An alternative approach, without loops"],"metadata":{"id":"v2UlKuYXE4hQ"}},{"cell_type":"code","source":["#the cumulative sum of chromosome lengths\n","chroms_cumulative = chroms.cumsum()\n","\n","#they need to be shifted of one\n","chroms_cumulative[1:23] = chroms_cumulative[0:22]\n","\n","#chrom 1 gets a zero shift\n","chroms_cumulative['chr01'] = 0\n","\n","#ready to map the values\n","df['SHIFT2'] = df['CHROM'].map(chroms_cumulative)\n","\n","#testing that we get the same results\n","all(df['SHIFT2'] == df['SHIFT'])"],"metadata":{"id":"ZGWLGFsMD1di"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["# SA5) Create new column X (POS + SHIFT)"],"metadata":{"id":"DWasAi-pSSvS"}},{"cell_type":"code","source":["df['X'] = df['POS'] + df['SHIFT']\n","\n","df[df['CHROM'] == 'chr02'].head()"],"metadata":{"id":"p8osIK_QSEQu"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["# SA6) Do the plot"],"metadata":{"id":"kKKeXJ2JSzWg"}},{"cell_type":"code","source":["plt.scatter(x=df['X'], y=df['SCORE'])"],"metadata":{"id":"vFUsncLRSnMd"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["# SA7) Define a chrom-color dictionary\n","\n","That is, a dictionary where chromosome names are keys and named colors are values. To obtain the same result as in the example shown at the beginning of this document just alternate \"red\" and \"blue\"."],"metadata":{"id":"KVVoNVq1TCmI"}},{"cell_type":"code","source":["TRUE_color = 'red'\n","FALSE_color = 'blue'\n","\n","colors = {}\n","toggle = True\n","\n","for c in chroms.index:\n","  if toggle:\n","    colors[c] = TRUE_color\n","  else:\n","    colors[c] = FALSE_color\n","  \n","  toggle = not toggle\n","\n","colors"],"metadata":{"id":"b4wzx7CdTW7M"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["# SA8) Redo the plot, using the color map defined\n","\n","You need to specify the color for every row in your dataframe. However we can use the dictonary we just defined above and the handy [.map()](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.Series.map.html) pandas function to do the job for us."],"metadata":{"id":"HOxBEcOHViWy"}},{"cell_type":"code","source":["plt.scatter(x=df['X'], y=df['SCORE'], c=df['CHROM'].map(colors))"],"metadata":{"id":"1FsLR1KoVWsC"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["# SA9) Compute, for each chromosome, the median point"],"metadata":{"id":"xYw1FFMPV6Sj"}},{"cell_type":"code","source":["medians = df.groupby('CHROM')['X'].median()\n","medians"],"metadata":{"id":"x1hOwF_rXZ2J"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["# SA10) Add a tick + label on the x axis in the middle of each chromosome\n","\n","You should look for a function that manages X ticks. I wonder how it could be called."],"metadata":{"id":"rC0wI8O7V_5N"}},{"cell_type":"code","source":["plt.scatter(x=df['X'], y=df['SCORE'], c=df['CHROM'].map(colors))\n","plt.xticks(ticks = medians, labels = medians.index)"],"metadata":{"id":"YObWDP4AXmAk"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["plt.figure(figsize=(12, 8))\n","plt.scatter(x=df['X'], y=df['SCORE'], c=df['CHROM'].map(colors))\n","plt.xticks(ticks = medians, labels = medians.index, rotation = -90)"],"metadata":{"id":"D8dorOOkX_Cc"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["# SA11) Put everything in a function\n","\n","Functions are handy."],"metadata":{"id":"Uyf4SEpMWODL"}},{"cell_type":"code","source":["# we expect the dataframe passed in to have three fields: CHROM, POS and SCORE\n","\n","def my_manhattan(data):\n","  #a local copy of the data, so we can add columns to\n","  #the dataframe without changing the original\n","  data = data.copy()\n","\n","  #sorting by chromosome\n","  data.sort_values(by = 'CHROM', inplace = True)\n","\n","  #length of each chromosome\n","  lengths = data.groupby('CHROM')['POS'].max()\n","\n","  #computing the chromosome shift\n","  data['SHIFT'] = 0\n","  shift = 0\n","  for c in lengths.index:\n","    sel = data['CHROM'] == c\n","    data.loc[sel, 'SHIFT'] = shift\n","    shift = shift + lengths[c]\n","  \n","  #computing the actual x values\n","  data['X'] = data['POS'] + data['SHIFT']\n","  \n","  #creating a color dictionary of alternating colors\n","  values = ['red', 'blue'] * int(len(lengths) / 2)\n","  if len(lengths) > len(values):\n","    values.append('red')\n","  colors = dict(zip(lengths.index, values))\n","\n","  #computing the medians\n","  medians = data.groupby('CHROM')['X'].median()\n","\n","  #doing the plot\n","  plt.figure(figsize=(12, 8))\n","  plt.scatter(x=data['X'], y=data['SCORE'], c=data['CHROM'].map(colors))\n","  plt.xticks(ticks = medians, labels = medians.index, rotation = -90)\n","\n","  return(None)\n","\n","my_manhattan(df)"],"metadata":{"id":"6ECrhq0ZtUhv"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["# SA12) Put the function in a module\n","\n","You'll need to do that from YOUR computer."],"metadata":{"id":"q8WUprulWS3Q"}},{"cell_type":"code","source":[],"metadata":{"id":"Dr9i25RxaWvk"},"execution_count":null,"outputs":[]}]}